### PR TITLE
Add TLSv1.3 support.

### LINKED ISSUE #13506: TLS1.3 support
version:
image
config:
image
image

Port detection status：
image
image

but, old version(3.4.10) is ok:
version:
image
config:
image
image
Port detection status：
image
image

### PR DESCRIPTION
**Pull request by tsaarni**
This PR adds optional TLS min/max protocol versions fields to `TLSInfo` struct, which is used in both server and client configs. Instead of hardcoding max version to TLSv1.2 it lets Go select the max version. This effectively enables TLSv1.3 support for etcd. The min/max values are also exposed via optional server flags and config file fields so that users can explicitly set the minimum and/or maximum TLS versions, if they want.

It seems that TLSv1.3 was previously supported (prior etcd 3.5) but it got disabled as a workaround to fix a test suite failure which appeared when Go 1.13 release enabled TLSv1.3 by default, see discussion https://github.com/etcd-io/etcd/pull/11110#issuecomment-528617457. In this PR, the failing test case is changed to set the max version to TLSv1.2. This approach allows testing ciphers with <TLSv1.3 without disabling TLSv1.3 altogether.

Related documentation change https://github.com/etcd-io/website/pull/641.

Fixes #13506

Signed-off-by: Tero Saarni <tero.saarni@est.tech>

### CODE REVIEW

**Comment by tsaarni** [comment]
Hi @serathius, please could you check this PR. Thank You!

**Review by serathius — APPROVED** [review]
LGTM, is the min/max tls config exposed in etcd server binary flags?
cc @ahrtr @ptabor for second opinion

**Comment by tsaarni** [comment]
> is the min/max tls config exposed in etcd server binary flags

Currently not, I targeted only to the minimum change. I can add server flags & config file options as well if requested.

**Comment by serathius** [comment]
> > is the min/max tls config exposed in etcd server binary flags
> 
> Currently not, I targeted only to the minimum change. I can add server flags & config file options as well if requested.

Sounds good

**Comment by tsaarni** [comment]
I've now added config options for etcd server binary and new tests. I see the docs are in website repo so a follow up PR is needed for the docs after confirming the proposed flags are OK.

File: server/embed/config.go
Comment (tuminoid) [inline server/embed/config.go:230]: Can we have this as `--tls-min-version` and `--tls-max-version` ?

File: client/pkg/tlsutil/versions.go
Comment (tuminoid) [inline client/pkg/tlsutil/versions.go:23]: Since etcd only supports TLS 1.2 pre-PR, does it make sense to add support for obsolete, insecure and disabled-by-default TLS 1.0 and TLS 1.1? Enabling TLS 1.3 would be sufficient in my opinion, and configuration would then allow a meaningful choice between forcing TLS 1.2 only or TLS 1.3 only in addition to default behavior of enabling both.

**Review by tuminoid — COMMENTED** [review]
Thanks @tsaarni for the PR! I added two comments, let me know what you think @serathius.

**Comment by tsaarni** [comment]
Thanks @tuminoid I agree with your review comments, so I went ahead and made those changes. Eagerly waiting for more reviews :-)

**Comment by serathius** [comment]
cc @ptabor

**Comment by tsaarni** [comment]
@serathius One test failed but could it be a flake https://github.com/etcd-io/etcd/actions/runs/3994448937/jobs/6852174459?  Would it be possible to trigger a re-run?

File: server/embed/config.go
Comment (ahrtr) [inline server/embed/config.go:232]:
```suggestion
	TlsMinVersion string `json:"tls-min-version"`
	// MaxVersion is the maximum accepted TLS version between client/server and peers.
	TlsMaxVersion string `json:"tls-max-version"`
```

File: server/etcdmain/config.go
Comment (ahrtr) [inline server/etcdmain/config.go:219]: Please keep the description consistent with the messages in `help.go`.  The flags are only used by etcdserver. Suggest to simplify it to something like "`Minimum TLS version supported by etcd. Possible values: TLS12, TLS13.`"

File: tests/integration/v3_tls_test.go
Comment (ahrtr) [inline tests/integration/v3_tls_test.go:82]: minor comment, it would be better to refactor this test case using table driven + golang subtest.

File: server/embed/config_test.go
Comment (ahrtr) [inline server/embed/config_test.go:494]: Note that it doesn't cover the case `err == nil && tt.expectError`.

```suggestion
			if (err != nil) != tt.expectError {
```

**Comment by ahrtr** [comment]
Thank you @tsaarni 

Overall looks good to me, with a couple of minor comments.

File: server/etcdmain/help.go
Comment (ahrtr) [inline server/etcdmain/help.go:205]: Please update the messages as well, to keep consistent with the message in config.go

**Review by ahrtr — APPROVED** [review]
LGTM

Please consider to use `github.com/stretchr/testify` to simplify the test cases if you want.

**Comment by tsaarni** [comment]
Thank you @ahrtr, great feedback! 

If it's OK, I will leave the tests as they are, but I also like `github.com/stretchr/testify` asserts better. When looking around the test files I touched, I thought standard `testing` package was preferred, so I used that mostly.

**Comment by tsaarni** [comment]
I've submitted also a proposal for the documentation change https://github.com/etcd-io/website/pull/641.

File: client/pkg/tlsutil/versions.go
Comment (serathius) [inline client/pkg/tlsutil/versions.go:23]: Please create a const for "TLS12" and "TLS13", best if we had type alias for it.

File: client/pkg/tlsutil/versions.go
Comment (serathius) [inline client/pkg/tlsutil/versions.go:26]:
```suggestion
	"TLS1.2": tls.VersionTLS12,
	"TLS1.3": tls.VersionTLS13,
```

File: server/embed/config.go
Comment (serathius) [inline server/embed/config.go:679]: This should be done during configuration validation stage, not during apply.

File: server/embed/etcd.go
Comment (serathius) [inline server/embed/etcd.go:619]: This side effect approach of generating TLSInfo is very error prone. Are we sure we called `updateMinMaxVersions` in all places that TLSInfo is changed?


Would really like to have this code cleaned up.

File: server/embed/config.go
Comment (serathius) [inline server/embed/config.go:690]: Can you add validation of min-tls and max-tls flags here?

**Comment by serathius** [comment]
Please add e2e tests for new etcdserver flags.

File: server/embed/config.go
Comment (serathius) [inline server/embed/config.go:680]: This validation doesn't check if cipher suites are configured. As cipher suits only work with TLS 1.2, we should not allow users to set them if maxTLS is not set to TLS1.2

File: server/embed/etcd.go
Comment (tsaarni) [reply server/embed/etcd.go:619]: I copied the approach from how the ciphers seemed to be generated to `TLSInfo` by similar `updateCipherSuites()` function (called [here](https://github.com/etcd-io/etcd/blob/main/server/embed/config.go#L932) and [here]( https://github.com/etcd-io/etcd/blob/3b612ce345753f97156833e20e93bf78b56d5267/server/embed/config.go#L951)). I'm little doubtful if I'm able to do bigger refactoring for `TLSInfo` generation. Maybe there is an approach to clean up but with minimal impact to existing code?

File: client/pkg/tlsutil/versions.go
Comment (serathius) [reply client/pkg/tlsutil/versions.go:26]: Let's use correct semantics names, we don't want to run into Windows 9 problems.

File: server/embed/config.go
Comment (tuminoid) [reply server/embed/config.go:680]: > This validation doesn't check if cipher suites are configured. As cipher suits only work with TLS 1.2, we should not allow users to set them if maxTLS is not set to TLS1.2

You must be able to set cipher suites for TLS 1.2 in case you have both TLS 1.2 and TLS 1.3 enabled (the default), so what you said does not make sense. What you probably wanted is failing validation if min-version and max-version is TLS 1.3.

Comment (tsaarni) [reply server/embed/config.go]: I added a check when TLSv1.3 is the only enabled version. 

It is not out of the question that there will be configurability in the future versions of Go and the check might need to be removed later. At this time, it is of course unknown how the possible configurability would look like.

File: server/embed/config.go
Comment (serathius) [reply server/embed/config.go:670]: panic or `cfg.logger.Fatal`.

File: server/etcdmain/config.go
Comment (serathius) [inline server/etcdmain/config.go:220]:
```suggestion
	fs.StringVar(&cfg.ec.TlsMaxVersion, "tls-max-version", string(tlsutil.TLSVersionDefault), "Maximum TLS version supported by etcd. Possible values: TLS1.2, TLS1.3 (empty will be auto-populated by Go).")
```

Comment (serathius) [inline server/etcdmain/config.go:220]:
```suggestion
	fs.StringVar(&cfg.ec.TlsMaxVersion, "tls-max-version", "", "Maximum TLS version supported by etcd. Possible values: TLS1.2, TLS1.3 (empty defers to Go version).")
```

File: server/embed/etcd.go
Comment (serathius) [reply server/embed/etcd.go]: Sounds good.

**Review by serathius — APPROVED** [review]

**Comment by tsaarni** [comment]
I have made the requested changes but tests failed, possibly flukes again?

**Pull request merged**
